crates/zir-codegen/src/testing.rs

```diff
@@ -0,0 +1,347 @@
+pub fn sig_void_to_i64() -> FunctionSignature {
+    FunctionSignature::new().with_return(TypeDesc::Int(64))
+}
+
+/// Signature for a function taking two i64 and returning i64.
+pub fn sig_i64_i64_to_i64() -> FunctionSignature {
+    FunctionSignature::new()
+        .with_param(TypeDesc::Int(64))
+        .with_param(TypeDesc::Int(64))
+        .with_return(TypeDesc::Int(64))
+}
+
+/// Compiles a MIR body to IR using any backend implementing [`CodegenBackend`].
+///
+/// This function is backend-agnostic - it works with any backend that implements
+/// the [`CodegenBackend`] trait.
+///
+/// # Arguments
+///
+/// * `backend` - A mutable reference to a [`CodegenBackend`] implementation
+/// * `body` - The MIR body to compile
+/// * `sig` - The function signature
+///
+/// # Returns
+///
+/// The textual IR representation on success, or an error.
+pub fn compile_to_ir_text(
+    backend: &mut dyn CodegenBackend,
+    body: &Body<'_>,
+    sig: FunctionSignature,
+) -> CodegenResult<String> {
+    let ir = backend.compile_to_ir(body, sig)?;
+    match ir {
+        IrOutput::Text(text) => Ok(text),
+        IrOutput::Binary(bytes) => Ok(format!("<binary {} bytes>", bytes.len())),
+    }
+}
+
+/// A test case for backend-agnostic codegen testing.
+///
+/// This struct encapsulates a test case that can be run against any
+/// [`CodegenBackend`] implementation.
+pub struct CodegenTestCase<'a> {
+    /// Name of the test case.
+    pub name: &'static str,
+    /// The MIR body to compile.
+    pub body: Body<'a>,
+    /// The function signature.
+    pub signature: FunctionSignature,
+}
+
+impl<'a> CodegenTestCase<'a> {
+    /// Creates a new test case.
+    pub fn new(name: &'static str, body: Body<'a>, signature: FunctionSignature) -> Self {
+        Self { name, body, signature }
+    }
+
+    /// Runs this test case against a backend.
+    ///
+    /// Returns the generated IR text.
+    pub fn run(&self, backend: &mut dyn CodegenBackend) -> CodegenResult<String> {
+        compile_to_ir_text(backend, &self.body, self.signature.clone())
+    }
+}
+
+/// Creates the standard set of test cases for codegen verification.
+///
+/// This returns a vector of test case builders that can be used with any
+/// backend implementing [`CodegenBackend`].
+///
+/// The test cases cover:
+/// - Constant value generation
+/// - Negative constant handling
+/// - Binary operations (add)
+/// - Control flow (branching)
+pub fn standard_test_cases<'a>(arena: &'a Arena<'a>) -> Vec<CodegenTestCase<'a>> {
+    vec![
+        CodegenTestCase::new("const_42", create_const_function(arena, 42), sig_void_to_i64()),
+        CodegenTestCase::new(
+            "const_negative",
+            create_const_function(arena, -123),
+            sig_void_to_i64(),
+        ),
+        CodegenTestCase::new("add_function", create_add_function(arena), sig_i64_i64_to_i64()),
+        CodegenTestCase::new("max_function", create_max_function(arena), sig_i64_i64_to_i64()),
+    ]
+}
```

Comment:
Improve this, make easy to define, maybe use files or macroses, or unified way to define clif test